src/txmempool.cpp

```diff
@@ -1245,3 +1648,251 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
+    for (auto chunkit = m_chunks.rbegin(); chunkit != m_chunks.rend(); ++chunkit) {
+        if (!chunkit->txs.empty()) return chunkit->txs.back();
+    }
+    // Unreachable
+    assert(false);
+}
+
+void Cluster::AddTransaction(const CTxMemPoolEntry& entry, bool sort)
+{
+    m_chunks.emplace_back(entry.GetModifiedFee(), entry.GetTxSize());
+    m_chunks.back().txs.emplace_back(entry);
+    entry.m_cluster = this;
+    ++m_tx_count;
+    m_tx_size += entry.GetTxSize();
+    if (sort) Sort();
+    return;
+}
+
+void Cluster::RemoveTransaction(const CTxMemPoolEntry& entry)
+{
+    m_chunks[entry.m_loc.first].txs.erase(entry.m_loc.second);
+
+    // Chunk (or cluster) may now be empty, but this will get cleaned up
+    // when the cluster is re-sorted (or when the cluster is deleted) Note:
+    // if we cleaned up empty chunks here, then this would break the
+    // locations of other entries in the cluster. Since we would like to be
+    // able to do multiple removals in a row and then clean up the sort, we
+    // can't clean up empty chunks here.
+    --m_tx_count;
+    m_tx_size -= entry.GetTxSize();
+    return;
+}
+
+void Cluster::RechunkFromLinearization(std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& txs, bool reassign_locations)
+{
+    m_chunks.clear();
+    m_tx_size = 0;
+
+    for (auto txentry : txs) {
+        m_chunks.emplace_back(txentry.get().GetModifiedFee(), txentry.get().GetTxSize());
+        m_chunks.back().txs.emplace_back(txentry);
+        while (m_chunks.size() >= 2) {
+            auto cur_iter = std::prev(m_chunks.end());
+            auto prev_iter = std::prev(cur_iter);
+            double feerate_prev = prev_iter->fee*cur_iter->size;
+            double feerate_cur = cur_iter->fee*prev_iter->size;
+            // We only combine chunks if the feerate would go up; if two
+            // chunks have equal feerate, we prefer to keep the smaller
+            // chunksize (which is generally better for both mining and
+            // eviction).
+            if (feerate_cur > feerate_prev) {
+                prev_iter->fee += cur_iter->fee;
+                prev_iter->size += cur_iter->size;
+                prev_iter->txs.splice(prev_iter->txs.end(), cur_iter->txs, cur_iter->txs.begin(), cur_iter->txs.end());
+                m_chunks.erase(cur_iter);
+            } else {
+                break;
+            }
+        }
+        m_tx_size += txentry.get().GetTxSize();
+    }
+
+    if (reassign_locations) {
+        // Update locations of all transactions
+        for (size_t i=0; i<m_chunks.size(); ++i) {
+            for (auto it = m_chunks[i].txs.begin(); it != m_chunks[i].txs.end(); ++it) {
+                it->get().m_loc = {i, it};
+            }
+        }
+    }
+}
+
+namespace {
+
+template <typename SetType>
+std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> InvokeSort(size_t tx_count, const std::vector<Cluster::Chunk>& chunks)
+{
+    std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> txs;
```

Comment:
Can you clarify what txs, orig_txs and cluster are for, and what the general strategy is here?